### PR TITLE
Development

### PR DESCRIPTION
**Pull request by ashishra0**
- User Authentication with username
- Artists Search
- LastFM API Service
- User Search History

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+class User < ApplicationRecord
+  USERNAME_FORMAT = /\A[a-zA-Z0-9]+\z/.freeze
+  has_secure_password
+  validates :password, presence: true
+  validates :username,
+            presence: true,
+            uniqueness: { case_sensitive: true },
+            length: { minimum: 5, maximum: 15 },
```
Comment (ankuriitk) [inline app/models/user.rb:8]: none of this is working - I was able to set a 3 char password

**Comment by ankuriitk** [comment]
@ashishra0 : nice test coverage. I noticed that your controllers are heavy. Ideally, controllers should have minimal code and models should be heavy. Small nit about the UI/UX: the history links are not clickable.

File: db/migrate/20190930164241_create_users.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
+class CreateUsers < ActiveRecord::Migration[5.2]
+  def change
+    create_table :users do |t|
+      t.string :username
+      t.string :password_digest
```
Comment (ankuriitk) [inline db/migrate/20190930164241_create_users.rb:5]: good that it is a digest

File: db/migrate/20191001154159_create_search_histories.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+class CreateSearchHistories < ActiveRecord::Migration[5.2]
+  def change
+    create_table :search_histories do |t|
+      t.references :user, foreign_key: true, index: true
+      t.string :keyword
```
Comment (ankuriitk) [inline db/migrate/20191001154159_create_search_histories.rb:5]: shouldn't it be citext?

Comment (ankuriitk) [reply db/migrate/20191001154159_create_search_histories.rb:5]: right now "kishore" and "Kishore" appear as separate results

Comment (ankuriitk) [reply db/migrate/20191001154159_create_search_histories.rb:5]: i guess thats ok - the history is a log of queries and not necessarily unique queries

**Comment by ankuriitk** [comment]
documentation is probably the other nit -> didn't find any documentation in code :)

File: app/models/user.rb (hunk 1 above)
Comment (ashishra0) [reply app/models/user.rb:8]: Oh, I did not add validation for password length. The username validations are working correct?

File: db/migrate/20190930164241_create_users.rb (hunk 2 above)
Comment (ashishra0) [reply db/migrate/20190930164241_create_users.rb:5]: yes, password is encrypted using that bcrypt gem

File: db/migrate/20191001154159_create_search_histories.rb (hunk 3 above)
Comment (ashishra0) [reply db/migrate/20191001154159_create_search_histories.rb:5]: I wasn't aware of citext. But now I am.

Comment (ashishra0) [reply db/migrate/20191001154159_create_search_histories.rb:5]: > right now "kishore" and "Kishore" appear as separate results
This shouldn't have happened. Apologies. I will fix this asap

**Comment by ashishra0** [comment]
> documentation is probably the other nit -> didn't find any documentation in code :)

I will take care of this. Thank you for reviewing.

**Pull request merged**